runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java

```diff
@@ -166,13 +167,13 @@ public Reader createReader(PipelineOptions options, Checkpoint<T> checkpoint)
     }
 
     @Override
-    public Coder<T> getDefaultOutputCoder() {
+    public Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
       return boundedSource.getDefaultOutputCoder();
     }
 
     @SuppressWarnings({"rawtypes", "unchecked"})
     @Override
-    public Coder<Checkpoint<T>> getCheckpointMarkCoder() {
+    public Coder<Checkpoint<T>> getCheckpointMarkCoder() throws CannotProvideCoderException {
```

Comment:
Hmm sorry, I just thought about this some more and realized that this is very problematic. Since now it is legal to have a source that does not provide a coder for itself - such a source can be used here too (in BoundedToUnboundedSourceAdapter). It looks like, in that case, getCheckpointMarkCoder() will fail - and this is a problem: when you just apply Read.from() to a source that doesn't provide a coder, you can simply set the coder explicitly on the resulting PCollection - but there's no such alternative for getCheckpointMarkCoder. And I don't think there should be - checkpoint mark coder of a source is an implementation detail of the source, so it's the source's responsibility to know this coder and the method probably should not be allowed to throw a CannotProvideCoderException.

What's worse - this class (UnboundedReadFromBoundedSource) is hidden from the user, and gets transparently applied by Dataflow runner at https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L365. That class (StreamingUnboundedRead) will call source.getDefaultOutputCoder() and will fail. Instead, it should infer the checkpoint mark coder for the BoundedToUnboundedSourceAdapter from the coder of the output PCollection - for that, StreamingUnboundedRead must stop being a ReflectiveRootOverrideFactory and should be an actual PTransformOverrideFactory and get the output PCollection from the AppliedPTransform.